File: typestate-builder-java/src/main/java/io/github/raboro/typestatebuilderjava/HouseBuilder.java
package io.github.raboro.typestatebuilderjava;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class HouseBuilder {
    private Optional<List<House.Wall>> walls;
    private Optional<List<House.Door>> doors;
    private Optional<House.Garage> garage;
    private Optional<House.Basement> basement;
    private Optional<List<House.Window>> windows;

    public HouseBuilder() {
        this.walls = Optional.empty();
        this.doors = Optional.empty();
        this.garage = Optional.empty();
        this.basement = Optional.empty();
        this.windows = Optional.empty();
    }

    public HouseBuilder setWalls(List<House.Wall> walls) {
        this.walls = Optional.of(walls);
        return this;
    }

    public HouseBuilder setDoors(List<House.Door> doors) {
        this.doors = Optional.of(doors);
        return this;
    }

    public HouseBuilder setGarage(House.Garage garage) {
        this.garage = Optional.of(garage);
        return this;
    }

    public HouseBuilder setBasement(House.Basement basement) {
        this.basement = Optional.of(basement);
        return this;
    }

    public HouseBuilder setWindows(List<House.Window> windows) {
        this.windows = Optional.of(windows);
        return this;
    }

    public House build() {
        return new House(
                this.walls.orElseGet(ArrayList::new),
                this.doors.orElseGet(ArrayList::new),
                this.garage,
                this.basement,
                this.windows
        );
    }
}

File: typestate-builder-ts/src/houseBuilder.ts
import { Wall, Door, Garage, Basement, _Window, House } from "./utils";

export class HouseBuilder {
    private walls?: Wall[];
    private doors?: Door[];
    private garage?: Garage;
    private basement?: Basement;
    private windows?: _Window[];

  static new(): HouseBuilder {
    return new HouseBuilder();
  }

  setWalls(walls: Wall[]): this {
    this.walls = walls;
    return this;
  }

  setDoors(doors: Door[]): this {
    this.doors = doors;
    return this;
  }

  setGarage(garage: Garage): this {
    this.garage = garage;
    return this;
  }

  setBasement(basement: Basement): this {
    this.basement = basement;
    return this;
  }

  setWindows(windows: _Window[]): this {
    this.windows = windows;
    return this;
  }

  build(): House {
    return new House(
      this.walls ?? [],
      this.doors ?? [],
      this.garage,
      this.basement,
      this.windows  
    );
  }

}
